alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/AbstractOrientingBehavior.kt

```diff
@@ -0,0 +1,318 @@
+     * suitability in order to reach the provided destination. Rankings should be in [1,N].
+     * A lower rank means the correspondent edge is preferable to reach the given destination.
+     *
+     * When following a route provided by the cognitive map, the agent often is in the following
+     * situation: he knows the spatial structure of the current room and the next sub-destination
+     * he want to reach, but has no other information regarding the spatial structure between
+     * his room and that position (remember that it is assumed the cognitive map provide only
+     * a boolean information regarding the connection of landmarks, no extra info is stored).
+     * This method should implement an algorithm allowing the pedestrian to perform an educated
+     * guess of which crossing to take in order to get closer to the provided destination.
+     */
+    protected abstract fun computeEdgeRankings(currRoom: N1, destination: P): Map<E1, Int>
+
+    /**
+     * Computes the next sub-destination the pedestrian will move towards, provided an edge the
+     * pedestrian wants to cross. The provided edge belongs to the room the agent is located into
+     * (i.e. it is in sight of the agent), the returned position must be in sight of him as well,
+     * which means no obstacle should be placed between such destination and his current
+     * position.
+     *
+     * This method is mainly about exploiting the extra data stored in the edge (for instance
+     * its shape and location in the room boundary) to determine which point the pedestrian
+     * shall point towards.
+     */
+    protected abstract fun computeSubdestination(targetEdge: E1): P
+
+    /**
+     * Move the pedestrian towards a position which is guaranteed to be in sight (i.e. no
+     * obstacle is placed between him and such position).
+     */
+    protected open fun moveTowards(currRoom: N1, target: P, currPosition: P) {
+        Seek(env, this, pedestrian, *target.cartesianCoordinates).execute()
+    }
+
+    /**
+     * Weights an edge.
+     */
+    protected open fun weight(e: E1, rank: Int?): Double {
+        val fVolatileMem = 2.0.pow(pedestrian.volatileMemory[e.to] ?: 0)
+        var fCognitiveMap = 1.0
+        rank?.let { fCognitiveMap -= 0.5.pow(rank) }
+        val fFinal = if (envGraph.containsDestination(e.to)) 0.1 else 1.0
+        //val fImpasse = if (impasses.contains(e.to)) 10.0 else 1.0
+        val fImpasse = if (isImpasse(e.to)) 10.0 else 1.0
+        return fVolatileMem * fCognitiveMap * fFinal * fImpasse
+    }
+
+    /*
+     * Registers visit in volatile memory.
+     */
+    private fun OrientingAgent<P, A, N2, *>.registerVisit(to: N1) {
+        volatileMemory[to] = (volatileMemory[to] ?: 0) + 1
+    }
+
+    private fun isImpasse(area: N1): Boolean =
+        pedestrian.volatileMemory.contains(area) &&
+            envGraph.edgesFrom(area).map { it.to }.distinct().count() <= 1
```

Comment:
what does it mean `it.to`? To what?

Reply:
`it` is an edge and `to` represents the node this edge points to. Maybe it creates some confusion in combo with `map`, does `map { edge -> edge.to }` looks better?

Reply:
Transform them as constructor parameters with that (documented) default

Reply:
uhmm... head/tail -- from https://en.wikipedia.org/wiki/Graph_theory